server/controllers/users.ts

```diff
@@ -179,6 +183,167 @@ export const signin = async (req, res, next) => {
   }
 };
 
+type SignupResquestSession = {
+  secret: string;
+  tries: number;
+  userId: number;
+};
+
+/**
+ * Creates a new User and sends a new OTP verification code.
+ */
+export const signup: express.RequestHandler<unknown, unknown, { email: string; password?: string }> =
+  async function signup(
+    req,
+    res,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    next,
+  ) {
+    const ipRateLimit = new RateLimit(`signup_ip_${req.ip}`, 3, 15 * 60);
+    if (!(await ipRateLimit.registerCall())) {
+      res.status(403).send({
+        error: { message: 'Rate limit exceeded' },
+      });
+      return;
+    }
+
+    const { email, password } = req.body;
+    const sanitizedEmail = email.toLowerCase();
+    const otpSessionKey = `otp_signup_${sanitizedEmail}`;
+
+    if (!isValidEmail(sanitizedEmail)) {
+      res.status(400).send({
+        error: { message: 'Invalid email address' },
+      });
+      return;
+    }
+    const emailRateLimit = new RateLimit(`signup_email_${sanitizedEmail}`, 3, 15 * 60);
+    if (!(await emailRateLimit.registerCall())) {
+      res.status(403).send({
+        error: { message: 'Rate limit exceeded' },
+      });
+      return;
+    }
+
+    // Check if OTP request already exists
+    const otpSession: SignupResquestSession = await sessionCache.get(otpSessionKey);
+    if (otpSession) {
+      res.status(401).send({
+        error: { message: 'OTP request already exists' },
+      });
+      return;
+    }
+
+    // Check if Users exists
+    let user = await models.User.findOne({ where: { email: sanitizedEmail } });
+    if (user) {
+      if (user.confirmedAt) {
+        res.status(403).send({
+          error: { message: 'User already exists' },
+        });
+        return;
+      }
+    } else {
+      user = await models.User.create({
+        email: sanitizedEmail,
+        confirmedAt: null,
+        data: {
+          creationRequest: {
+            ip: req.ip,
+            userAgent: req.header('user-agent'),
+          },
+        },
+      });
+      if (password) {
+        await user.setPassword(password);
+      }
+    }
+
+    const otp = randomInt(100000, 999999).toString();
+    const secret = await bcrypt.hash(otp, 10);
+    const session: SignupResquestSession = {
+      secret,
+      tries: 0,
+      userId: user.id,
+    };
+    await sessionCache.set(otpSessionKey, session, 5 * 60); // 5 minutes
```

Comment:
- 5 minutes seems very short to confirm an email, can we extend to 30min?
- I'd suggest putting this duration in `server/lib/auth.ts`, along token expirations durations, and reusing that in the email template to make sure we use the same value everywhere.